cpp/arcticdb/storage/storage_credential.hpp

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <azure/identity/default_azure_credential.hpp>
+#include <arcticdb/util/variant.hpp>
+
+namespace arcticdb::storage {
+
+    using AzureBaseCredential = Azure::Core::Credentials::TokenCredential;
+    using VariantStorageCredential = std::variant<std::monostate, std::shared_ptr<AzureBaseCredential>>;
+    class StorageCredential {
+    public:
+        VariantStorageCredential storage_credential_;
+        
+        const VariantStorageCredential& variant() const {
+            return storage_credential_;
+        }
+
+        void set_azure_credential(std::shared_ptr<AzureBaseCredential>& storage_credential) {
+            storage_credential_ = storage_credential;
+        }
+
+        [[nodiscard]] std::shared_ptr<AzureBaseCredential> get_azure_credential() const{
+            return util::variant_match(storage_credential_, 
+                [](const std::shared_ptr<AzureBaseCredential>& storage_credential) {
+                    return storage_credential;
+                },
+                [](std::monostate) {
+                    return std::shared_ptr<AzureBaseCredential>();
```

Comment:
The variant pattern used for `VariantStorageOverride` seems applicable here. But I think we've ended up in a bit of a halfway house with this function `get_azure_credential`. By the time we have that, this might as well be a struct. Instead we should expose some visitation pattern like in `apply_storage_override` so that consumers reading the override are agnostic as to the type?